prusti-specs/src/rewriter.rs

```diff
@@ -209,6 +223,159 @@ impl AstRewriter {
+                || -> bool {
+                    #expr
+                };
+            }
+        })
+    }
+}
+
+pub fn translate_empty_assertion() -> syn::Expr {
+    parse_quote_spanned! {
+        Span::call_site() =>
+        true
+    }
+}
+
+pub fn translate_pledge_rhs_only(reference: Option<syn::Expr>, rhs: syn::Expr) -> syn::Expr {
+    if let Some(reference) = reference {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge_rhs(#reference, #rhs)
+        }
+    } else {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge_rhs(result, #rhs)
+        }
+    }
+}
+
+pub fn translate_pledge(reference: Option<syn::Expr>, lhs: syn::Expr, rhs: syn::Expr) -> syn::Expr {
+    if let Some(reference) = reference {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge(#reference, #lhs, #rhs)
+        }
+    } else {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge(result, #lhs, #rhs)
+        }
+    }
+}
+
+pub fn translate_implication(lhs: syn::Expr, rhs: syn::Expr) -> syn::Expr {
+    parse_quote_spanned! {
+        lhs.span().join(rhs.span()).unwrap() =>
+        implication(#lhs, #rhs)
+    }
+}
+
+pub fn translate_conjunction(mut conjuncts: Vec<syn::Expr>) -> syn::Expr {
+    debug_assert!(conjuncts.len() != 0, "empty conjuncts given");
+    if conjuncts.len() == 1 {
+        conjuncts.pop().unwrap()
+    } else {
+        let last = conjuncts.pop().unwrap();
+        let rest = translate_conjunction(conjuncts);
+        parse_quote_spanned! {
+            rest.span().join(last.span()).unwrap() =>
+            #rest && #last
+        }
+    }
+}
+
+fn args_to_tokens(mut args: Vec<(syn::Ident, syn::Type)>) -> TokenStream {
+    if args.len() == 0 {
+        TokenStream::new()
+    } else if args.len() == 1 {
+        let (ident, typ) = args.pop().unwrap();
+        quote_spanned! {
+            ident.span().join(typ.span()).unwrap() =>
+            #ident: #typ
+        }
+    } else {
+        let (ident, typ) = args.pop().unwrap();
+        let rest = args_to_tokens(args);
+        quote_spanned! {
+            rest.span().join(typ.span()).unwrap() =>
+            #rest, #ident: #typ
+        }
+    }
+}
+
+pub fn translate_spec_entailment(closure: syn::Expr, args: Vec<(syn::Ident, syn::Type)>, pres: Vec<syn::Expr>, posts: Vec<syn::Expr>) -> syn::Expr {
+    let arg_tokens = args_to_tokens(args);
+    let pre_conjuncts = translate_conjunction(pres);
+    let post_conjuncts = translate_conjunction(posts);
+    parse_quote_spanned! {
+        // TODO: get the right span
+        Span::call_site() =>
+        entailment(#closure, |#arg_tokens| {
+            {
+                #pre_conjuncts
+            } &&
+            {
+                #post_conjuncts
```

Comment:
These (pre and post) should probably be separate arguments to `entailment` under separate closures. `post` should be able to use `result` (or `cl_result`), `pre` should not.

Reply:
Partially fixed by 33723fd23eaa1171543ec8351422a3dcf5e66c56'

I'm not quite sure how to get the type of the result here.

Reply:
Does something like 8dc5deb602a9a11d28486b914bf81b1185b64e30 work?

Reply:
No, `: T` like that would just trigger a typing error, it does not refer to the `T` in `entailment` (which you can remove again). Actually it turns out we can't know the result type without having the user declare it explicitly. I've made changes to the parser in the closure branch but I guess these are not on `master` yet. You can assume that the type is either known and is passed to this method somehow, or just put `i32` for now, which is what the behaviour was on `master` IIRC.

Reply:
I've made it an `i32` for now in 81ebcf180f7230702581aafa798071b09b8bf5ee.